web/src/main/java/com/futurewei/alcor/web/entity/dataplane/InternalPortEntity.java

```diff
@@ -27,12 +27,15 @@
     @JsonProperty("routes")
     private List<RouteEntity> routes;
 
-    @JsonProperty("neighbor_info")
-    private List<NeighborInfo> neighborInfos;
-
     @JsonProperty("binding_host_ip")
     private String bindingHostIP;
 
+    @JsonProperty("connected_router_id")
+    private String connectedRouterId;
+
+    @JsonProperty("neighbor_info")
```

Comment:
@chenpiaoping @haboy52581 

I am a bit confused here https://github.com/futurewei-cloud/alcor/pull/364/commits/8396de359288628c81ebac0853c96718039358a8 first changed neighbor_info to neighbor_ids, which were what we discussed and closed in this week's meeting. 

Not sure why this field comes back to neighbor_info. Should we change to neighbor_ids?

Reply:
[@chenpiaoping on Slack]
"I removed the neighbor field from InternalPortEntity, I don't think it needs this field.

Only left the neighbor table in NetworkConfiguration. Do you think this is all right?"

[@xieus] I had misunderstanding of your message on Slack. I thought you talked about "replace neighbor_info with neighbor_ids" but didn't realize you actually meant "removing neighbor field completely".
 
@chenpiaoping I think neighbor_ids are still needed. Think about this scenario:

Existing ports: P1 in Subnet1, P2 in Subnet2, P3 in Subnet3
Add two new ports: P4 in subnet1 and P5 in subnet2 

P1 is P4's L2 neighbor and P5's L3 neighbor
P2 is P4's L3 neighbor and P5's L2 neighbor
P3 is the L3 neighbor of both P4 and P5

In the message from PM to DPM:

New ports: [P4, P5]
Neighbor ports: [{P1, L2}, {P2, L3}, {P3, L3}, {P1, L3}, {P2, L2}]

The neighbor_ids field in InternalPortEntity would help DPM to narrow down its neighbor list.

Reply:
@xieus  Please review this commit: https://github.com/futurewei-cloud/alcor/pull/364/commits/5d7402d569c918b758f50a93621fb5ac7ccbca23

Reply:
My second comments was about https://github.com/futurewei-cloud/alcor/commit/5d7402d569c918b758f50a93621fb5ac7ccbca23. I think our messages were just crossed each other. Please take a look at my second comment.

Reply:
oh, Maybe it's because Xiaodong also modified it on this PR, and he also added some other fields to the InternalPortEntity.

Reply:
PM put all the neighbors together and call it the neighbor table. When a subnet connects to the router, PM set the type of all neighbors in this subnet to L3 neighbor and InternalPortEntity gets all neighbors of type L3 in the neighbor table as its L3 neighbors(excluding neighbors with the same subnet_id) . Get all the neighbors in the neighbor table that have the same subnet_id as L2 neighbors.
When creating / deleting a gateway interface, PM sends DPM the latest neighbor table

Reply:
Neighbor table:
subnet1: [P1, P4]
subnet2: [P2, P5]
subnet3: [P3]

Suppose subnet1 connects to the router，DPM creates the following neighbor relationships based on the neighbor table:
P1's L2 neighbro: [P4], P1's L3 neighbro: []
P2's L2 neighbro: [P5], P1's L3 neighbro: [P1, P4]
P3's L2 neighbro: [], P1's L3 neighbro: [P1, P4]
P4's L2 neighbro: [P1], P1's L3 neighbro: []
P5's L2 neighbro: [P2], P1's L3 neighbro: [P1, P4]

Reply:
Of course, all neighbor relationships can also be generated in PM, but more data needs to be sent to DPM, and the logic of generating neighbor relationships based on neighbor tables is simple and fast.

Reply:
@xieus  @haboy52581 The comment above is my idea，please review.

Reply:
On the basis of the above, suppose that P6 is created in subnet3 at this time, and the neighbor table that PM sends to DPM is as follows:
subnet1: [P1, P4]
subnet3: [P3, P6]

InternalPortEntity: P6
DPM creates the following neighbor relationships based on the neighbor table:
P6's L2 neighbros: [P3], P6's L3 neighbros: [P1, P4]

Reply:
> Neighbor table:
> subnet1: [P1, P4]
> subnet2: [P2, P5]
> subnet3: [P3]
> 
> Suppose subnet1 connects to the router，then The neighborType of P1 and P4 will be set to L3.
> DPM creates the following neighbor relationships based on the neighbor table:
> P1's L2 neighbors: [P4], P1's L3 neighbors: []
> P2's L2 neighbors: [P5], P2's L3 neighbors: [P1, P4]
> P3's L2 neighbors: [], P3's L3 neighbors: [P1, P4]
> P4's L2 neighbors: [P1], P4's L3 neighbors: []
> P5's L2 neighbors: [P2], P5's L3 neighbors: [P1, P4]

@chenpiaoping In the case listed above, are subnet2 and subnet3 connected to the router before subnet1 is connected to the router?
- If yes, L3 neighbors of P1 and P4 are [P2, P3, P5]
- If not, L3 neighbors of P2, P3 and P5 should be [] as subnet2/3 are not connected to subnet1

Trying to understand how the neighbor table is generated on the fly at DPM..

Reply:
Oh yes, you're right. the L3 neighbor relationships above is wrong, the idea here is generating the neighbor relationships based on the neighbor table.

Reply:
Sure no problem. I understand your motivation of reducing the "redundant" information passed from PM to DPM. @chenpiaoping 

The trick here is if DPM generates the neighbor relationship based on the neighbor table of PM, then the neighbor relationship must be a FULL relationship in that VPC or network at that specific movement (i.e., a snapshot), as DPM is stateless as of now. I am not saying that FULL is wrong, just that the DPM-ACA channel will be expected to be very busy, especially in the large VPC scenario.

Let us continue to take the above example and assume that subnet2 and subnet3 are connected to the router before subnet1 is connected to the router. 

DPM to ACA messages will be something like
P1's L2 neighbors: [P4], P1's L3 neighbors: [P2, P3, P5]
P2's L2 neighbors: [P5], P2's L3 neighbors: [P1, P4]
P3's L2 neighbors: [], P3's L3 neighbors: [P1, P4]
P4's L2 neighbors: [P1], P4's L3 neighbors: [P2, P3, P5]
P5's L2 neighbors: [P2], P5's L3 neighbors: [P1, P4]

This means that we are almost the same of Neutron. If we generate the neighbor list on PM, we may be able to do it better by only sending incremental update (i.e., DELTA type) to from DPM to ACA.

P1's L2 neighbors: [P4], P1's L3 neighbors: [P2, P3, P5]
P2's L2 neighbors: [] (no change), P2's L3 NEW neighbors: [P1, P4]
P3's L2 neighbors: [] (no change), P3's L3 NEW neighbors: [P1, P4]
P4's L2 neighbors: [P1], P4's L3 neighbors: [P2, P3, P5]
P5's L2 neighbors: [] (no change), P5's L3 NEW neighbors: [P1, P4]

If Subnet2 and Subnet3 are huge, then the message size reduction could be significant. 

Let me know what you think of.

Reply:
Yes, you are right, but as you mentioned above, if DPM sends neighbor relationships to DPM , neighbor relationships:
P1's L2 neighbors: [P4], P1's L3 neighbors: [P2, P3, P5].
P2's L3 NEW neighbors: [P1, P4].
P3's L3 NEW neighbors: [P1, P4].
P4's L2 neighbors: [P1], P4's L3 neighbors: [P2, P3, P5].
P5's L3 NEW neighbors: [P1, P4].
More redundant data is needed than sending neighbor tables,  if subnet is large this redundancy will be large(e.g.: all the subnet are connected to a router and add a port to one subnet). does DPM need database?

Reply:
Let me see if there's any better idea.

Reply:
Agreed that PM would send more data to DPM compared to a neighbor table. Some contract changes proposed in this PR by you and Xiaodong, actually reduce some of the redundancy, for example, moving the neighbor ports as the top-level resources in NetworkConfiguration class, compared to previously as part of new ports.

Therefore, PM to DPM:
```
New ports (full detail): 
[
	{P1 (full info) + neighbor_ids=[p2, p3, p5]},
	{P4 (full info) + neighbor_ids=[p2, p3, p5]}
]
Neighbor ports
[ {P2}, {P3}, {P5} ]
```

DPM to ACA:
```
P1's L2 neighbors: [P4], P1's L3 neighbors: [P2, P3, P5].
P2's L3 NEW neighbors: [P1, P4].
P3's L3 NEW neighbors: [P1, P4].
P4's L2 neighbors: [P1], P4's L3 neighbors: [P2, P3, P5].
P5's L3 NEW neighbors: [P1, P4]
```

Not that bad, right? :-)

Reply:
public class NeighborEntry {
    @JsonProperty("neighbor_type")
    private NeighborType neighborType;

    @JsonProperty("local_ip")
    private String localIp;   //use port ip as neighbor id

    @JsonProperty("neighbor_ip")
    private String neighborIp;
}

public class NetworkConfiguration {
  @JsonProperty("neighbor_info")
  private List<NeighborInfo> neighborInfos;

  @JsonProperty("neighbor_table")
  private List<NeighborEntry> neighborTable;
  ......
}

Reply:
NeighborEntry represents a neighbor relationship , and neighborTable records all neighbor relationships.

Reply:
If NeighborEntry represents A's neighbor is B, then there must be a NeighborEntry represents B's neighbor is A. To reduce redundant data, only one of them will remain in neighborTable.

Reply:
> Of course, all neighbor relationships can also be generated in PM, but more data needs to be sent to DPM, and the logic of generating neighbor relationships based on neighbor tables is simple and fast.

can PM do this? since PM have enough data

Reply:
> Yes, you are right, but as you mentioned above, if DPM sends neighbor relationships to DPM , neighbor relationships:
> P1's L2 neighbors: [P4], P1's L3 neighbors: [P2, P3, P5].
> P2's L3 NEW neighbors: [P1, P4].
> P3's L3 NEW neighbors: [P1, P4].
> P4's L2 neighbors: [P1], P4's L3 neighbors: [P2, P3, P5].
> P5's L3 NEW neighbors: [P1, P4].
> More redundant data is needed than sending neighbor tables, if subnet is large this redundancy will be large(e.g.: all the subnet are connected to a router and add a port to one subnet). does DPM need database?

why P2's L3 new neighbor is p1 and p4?

Reply:
> ```
> public class NeighborEntry {
>     @JsonProperty("neighbor_type")
>     private NeighborType neighborType;
> 
>     @JsonProperty("local_ip")
>     private String localIp;   //use port ip as neighbor id
> 
>     @JsonProperty("neighbor_ip")
>     private String neighborIp;
> }
> 
> public class NetworkConfiguration {
>   @JsonProperty("neighbor_info")
>   private List<NeighborInfo> neighborInfos;
> 
>   @JsonProperty("neighbor_table")
>   private List<NeighborEntry> neighborTable;
>   ......
> }
> ```

 @chenpiaoping  so in your case neighbor_table field would be passed to include all neighbors (for which scope, L2, L3, host level?) please elaborate more, thanks!